# IMPORT LIBRARY

In [ ]:
import numpy as np
import pandas as pd
from itertools import product
import scipy.stats as stats
import matplotlib.pyplot as plt
import seaborn as sns
import random
import itertools
from tqdm import tqdm
from scipy.stats import gmean


# DATA LOAD

In [3]:
df = pd.read_csv("../data/for_simulation.csv")

# 1. 최소 리워드를 제공 할 때 순이익이 0을 넘길 수 있는 성장율
- ('High': 3, 'Mid': 2, 'Low': 1, 'Very Low': 0)

In [8]:
increase_rate_combinations = []
reward_combinations = [
    {'High': 3, 'Mid': 2, 'Low': 1, 'Very Low': 0}]

for i in np.arange(5, 26, 0.1):
    increase_rate_combinations.append(i * 0.01)

all_combinations = list(product(reward_combinations, increase_rate_combinations))
# 시뮬레이션 결과 저장
simulation_results = []

# 모든 조합에 대해 시뮬레이션 실행

for reward_amounts, increase_rates in all_combinations:
    # Trust level별 초기 총 수익(판매 대금의 10%) 계산
    initial_revenues = {level: int(df[df['trust_level'] == level]['payment_value'].sum() * 0.1) for level in df['trust_level'].unique()}
    initial_trust_distribution = df['trust_level'].value_counts(normalize=True).to_dict()
    
    # 초기 수익 및 시뮬레이션 계산
    initial_total_revenue = sum(initial_revenues.values())
    reward_provided_total = 0
    simulation_total_revenue = 0

    for trust_level, reward in reward_amounts.items():
        # 리워드 제공액 계산
        trust_level_count = len(df[df['trust_level'] == trust_level])
        reward_provided = reward * trust_level_count
        reward_provided_total += reward_provided
        
        # 구매 증가율에 따른 시뮬레이션 수익
        initial_revenue = initial_revenues.get(trust_level, 0)

        simulation_revenue = int(initial_revenue * (1 + increase_rates))
        simulation_total_revenue += simulation_revenue
    
    # 시뮬레이션 순수익
    simulation_net_profit = simulation_total_revenue - reward_provided_total
    
    # 리워드 제공에 따른 추가 수익
    additional_revenue = simulation_net_profit - initial_total_revenue
    
    # 시뮬레이션 결과 저장
    simulation_results.append({
        'Reward Combination': reward_amounts,
        'Increase Rate Combination': round(increase_rates*100,1),
        'Initial Revenue': initial_total_revenue,
        'Simulation Revenue': simulation_total_revenue,
        'Reward Provided': reward_provided_total,
        'Simulation Net Profit': simulation_net_profit,
        'Additional Revenue': additional_revenue
    })




In [9]:
# 결과 DataFrame 생성
minreward_simulation = pd.DataFrame(simulation_results)

# 상위 10개의 시뮬레이션 결과 출력
print("다양한 시뮬레이션 결과:")

minreward_simulation[minreward_simulation['Additional Revenue']>0].sort_values(by=["Additional Revenue"],ascending=True)

다양한 시뮬레이션 결과:


Reward Combination  \
44   {'High': 3, 'Mid': 2, 'Low': 1, 'Very Low': 0}   
45   {'High': 3, 'Mid': 2, 'Low': 1, 'Very Low': 0}   
46   {'High': 3, 'Mid': 2, 'Low': 1, 'Very Low': 0}   
47   {'High': 3, 'Mid': 2, 'Low': 1, 'Very Low': 0}   
48   {'High': 3, 'Mid': 2, 'Low': 1, 'Very Low': 0}   
..                                              ...   
205  {'High': 3, 'Mid': 2, 'Low': 1, 'Very Low': 0}   
206  {'High': 3, 'Mid': 2, 'Low': 1, 'Very Low': 0}   
207  {'High': 3, 'Mid': 2, 'Low': 1, 'Very Low': 0}   
208  {'High': 3, 'Mid': 2, 'Low': 1, 'Very Low': 0}   
209  {'High': 3, 'Mid': 2, 'Low': 1, 'Very Low': 0}   

     Increase Rate Combination  Initial Revenue  Simulation Revenue  \
44                         9.4          1993048             2180392   
45                         9.5          1993048             2182386   
46                         9.6          1993048             2184379   
47                         9.7          1993048             2186371   
48                         9.8          1993048             2188365   
..                         ...              ...                 ...   
205                       25.5          1993048             2501273   
206                       25.6          1993048             2503267   
207                       25.7          1993048             2505259   
208                       25.8          1993048             2507252   
209                       25.9          1993048             2509245   

     Reward Provided  Simulation Net Profit  Additional Revenue  
44            186267                1994125                1077  
45            186267                1996119                3071  
46            186267                1998112                5064  
47            186267                2000104                7056  
48            186267                2002098                9050  
..               ...                    ...                 ...  
205           186267                2315006              321958  
206           186267                2317000              323952  
207           186267                2318992              325944  
208           186267                2320985              327937  
209           186267                2322978              329930  

[166 rows x 7 columns]

In [12]:
low_rate = minreward_simulation[minreward_simulation['Additional Revenue']>0].sort_values(by=["Additional Revenue"],ascending=True).reset_index(drop=True).loc[0,'Increase Rate Combination']
low_rate

9.4

9.4%의 상승을 해야 흑자

# 2. 최대 리워드를 제공 할 때 순이익이 0을 넘길 수 있는 성장율
- ('High': 3, 'Mid': 2, 'Low': 1, 'Very Low': 0)

In [20]:
increase_rate_combinations = []
reward_combinations = [
    {'High': 9, 'Mid': 6, 'Low': 4, 'Very Low': 0}]

for i in np.arange(5, 31, 0.1):
    increase_rate_combinations.append(i * 0.01)

all_combinations = list(product(reward_combinations, increase_rate_combinations))
# 시뮬레이션 결과 저장
simulation_results = []

# 모든 조합에 대해 시뮬레이션 실행

for reward_amounts, increase_rates in all_combinations:
    # Trust level별 초기 총 수익(판매 대금의 10%) 계산
    initial_revenues = {level: int(df[df['trust_level'] == level]['payment_value'].sum() * 0.1) for level in df['trust_level'].unique()}
    initial_trust_distribution = df['trust_level'].value_counts(normalize=True).to_dict()
    
    # 초기 수익 및 시뮬레이션 계산
    initial_total_revenue = sum(initial_revenues.values())
    reward_provided_total = 0
    simulation_total_revenue = 0

    for trust_level, reward in reward_amounts.items():
        # 리워드 제공액 계산
        trust_level_count = len(df[df['trust_level'] == trust_level])
        reward_provided = reward * trust_level_count
        reward_provided_total += reward_provided
        
        # 구매 증가율에 따른 시뮬레이션 수익
        initial_revenue = initial_revenues.get(trust_level, 0)

        simulation_revenue = int(initial_revenue * (1 + increase_rates))
        simulation_total_revenue += simulation_revenue
    
    # 시뮬레이션 순수익
    simulation_net_profit = simulation_total_revenue - reward_provided_total
    
    # 리워드 제공에 따른 추가 수익
    additional_revenue = simulation_net_profit - initial_total_revenue
    
    # 시뮬레이션 결과 저장
    simulation_results.append({
        'Reward Combination': reward_amounts,
        'Increase Rate Combination': round(increase_rates*100,1),
        'Initial Revenue': initial_total_revenue,
        'Simulation Revenue': simulation_total_revenue,
        'Reward Provided': reward_provided_total,
        'Simulation Net Profit': simulation_net_profit,
        'Additional Revenue': additional_revenue
    })




In [21]:
# 결과 DataFrame 생성
maxreward_simulation = pd.DataFrame(simulation_results)

# 상위 10개의 시뮬레이션 결과 출력
print("다양한 시뮬레이션 결과:")

maxreward_simulation[maxreward_simulation['Additional Revenue']>0].sort_values(by=["Additional Revenue"],ascending=True)

다양한 시뮬레이션 결과:


Reward Combination  \
242  {'High': 9, 'Mid': 6, 'Low': 4, 'Very Low': 0}   
243  {'High': 9, 'Mid': 6, 'Low': 4, 'Very Low': 0}   
244  {'High': 9, 'Mid': 6, 'Low': 4, 'Very Low': 0}   
245  {'High': 9, 'Mid': 6, 'Low': 4, 'Very Low': 0}   
246  {'High': 9, 'Mid': 6, 'Low': 4, 'Very Low': 0}   
247  {'High': 9, 'Mid': 6, 'Low': 4, 'Very Low': 0}   
248  {'High': 9, 'Mid': 6, 'Low': 4, 'Very Low': 0}   
249  {'High': 9, 'Mid': 6, 'Low': 4, 'Very Low': 0}   
250  {'High': 9, 'Mid': 6, 'Low': 4, 'Very Low': 0}   
251  {'High': 9, 'Mid': 6, 'Low': 4, 'Very Low': 0}   
252  {'High': 9, 'Mid': 6, 'Low': 4, 'Very Low': 0}   
253  {'High': 9, 'Mid': 6, 'Low': 4, 'Very Low': 0}   
254  {'High': 9, 'Mid': 6, 'Low': 4, 'Very Low': 0}   
255  {'High': 9, 'Mid': 6, 'Low': 4, 'Very Low': 0}   
256  {'High': 9, 'Mid': 6, 'Low': 4, 'Very Low': 0}   
257  {'High': 9, 'Mid': 6, 'Low': 4, 'Very Low': 0}   
258  {'High': 9, 'Mid': 6, 'Low': 4, 'Very Low': 0}   
259  {'High': 9, 'Mid': 6, 'Low': 4, 'Very Low': 0}   

     Increase Rate Combination  Initial Revenue  Simulation Revenue  \
242                       29.2          1993048             2575016   
243                       29.3          1993048             2577009   
244                       29.4          1993048             2579003   
245                       29.5          1993048             2580994   
246                       29.6          1993048             2582988   
247                       29.7          1993048             2584982   
248                       29.8          1993048             2586975   
249                       29.9          1993048             2588967   
250                       30.0          1993048             2590959   
251                       30.1          1993048             2592954   
252                       30.2          1993048             2594947   
253                       30.3          1993048             2596939   
254                       30.4          1993048             2598932   
255                       30.5          1993048             2600926   
256                       30.6          1993048             2602919   
257                       30.7          1993048             2604911   
258                       30.8          1993048             2606904   
259                       30.9          1993048             2608899   

     Reward Provided  Simulation Net Profit  Additional Revenue  
242           580419                1994597                1549  
243           580419                1996590                3542  
244           580419                1998584                5536  
245           580419                2000575                7527  
246           580419                2002569                9521  
247           580419                2004563               11515  
248           580419                2006556               13508  
249           580419                2008548               15500  
250           580419                2010540               17492  
251           580419                2012535               19487  
252           580419                2014528               21480  
253           580419                2016520               23472  
254           580419                2018513               25465  
255           580419                2020507               27459  
256           580419                2022500               29452  
257           580419                2024492               31444  
258           580419                2026485               33437  
259           580419                2028480               35432

In [22]:
high_rate = maxreward_simulation[maxreward_simulation['Additional Revenue']>0].sort_values(by=["Additional Revenue"],ascending=True).reset_index(drop=True).loc[0,'Increase Rate Combination']
high_rate

29.2

In [24]:

inbcrease_rate = round(gmean([low_rate, high_rate]),1)
inbcrease_rate

16.6

> Increase_rate = 16.6

# 3. 리워드 추출
- 리워드를 제공하면 Increase rate의 성장을 이룬다고 가정
- 이때, 기존 데이터를 활용하여 |수익| 을 오름차순 정렬하여 리워드 조합 8개 추출

In [25]:
reward_combinations = []

very_low_reward = 0  # 디폴트 0
reward_min = 3
reward_max = 9
for i in range(reward_min, reward_max+1):
    # 랜덤 리워드 금액 설정
    high_reward = i
    for j in range(2,int(round(i*2/3))+1):
        mid_reward = j
        for k in range(1,int(round(j*2/3))+1):
            low_reward = k
            reward_combinations.append({
                'High': high_reward,
                'Mid': mid_reward,
                'Low': low_reward,
                'Very Low': very_low_reward
            })

increase_rate_combinations = [inbcrease_rate * 0.01]

all_combinations = list(product(reward_combinations, increase_rate_combinations))
# 시뮬레이션 결과 저장
simulation_results = []

# 모든 조합에 대해 시뮬레이션 실행

for reward_amounts, increase_rates in all_combinations:
    # Trust level별 초기 총 수익(판매 대금의 10%) 계산
    initial_revenues = {level: int(df[df['trust_level'] == level]['payment_value'].sum() * 0.1) for level in df['trust_level'].unique()}
    initial_trust_distribution = df['trust_level'].value_counts(normalize=True).to_dict()
    
    # 초기 수익 및 시뮬레이션 계산
    initial_total_revenue = sum(initial_revenues.values())
    reward_provided_total = 0
    simulation_total_revenue = 0

    for trust_level, reward in reward_amounts.items():
        # 리워드 제공액 계산
        trust_level_count = len(df[df['trust_level'] == trust_level])
        reward_provided = reward * trust_level_count
        reward_provided_total += reward_provided
        
        # 구매 증가율에 따른 시뮬레이션 수익
        initial_revenue = initial_revenues.get(trust_level, 0)

        simulation_revenue = int(initial_revenue * (1 + increase_rates))
        simulation_total_revenue += simulation_revenue
    
    # 시뮬레이션 순수익
    simulation_net_profit = simulation_total_revenue - reward_provided_total
    
    # 리워드 제공에 따른 추가 수익
    additional_revenue = simulation_net_profit - initial_total_revenue
    
    # 시뮬레이션 결과 저장
    simulation_results.append({
        'Reward Combination': reward_amounts,
        'Increase Rate Combination': round(increase_rates*100,1),
        'Initial Revenue': initial_total_revenue,
        'Simulation Revenue': simulation_total_revenue,
        'Reward Provided': reward_provided_total,
        'Simulation Net Profit': simulation_net_profit,
        'Additional Revenue': additional_revenue
    })




In [28]:
# 결과 DataFrame 생성
reward_combinations_simulation = pd.DataFrame(simulation_results)

# 상위 10개의 시뮬레이션 결과 출력
print("다양한 시뮬레이션 결과:")
reward_combinations_simulation['distance_from_zero'] = abs(reward_combinations_simulation['Additional Revenue'])
reward_combinations_simulation.sort_values(by=["distance_from_zero"],ascending=True)

다양한 시뮬레이션 결과:


Reward Combination  Increase Rate Combination  \
10  {'High': 6, 'Mid': 4, 'Low': 1, 'Very Low': 0}                       16.6   
16  {'High': 7, 'Mid': 4, 'Low': 1, 'Very Low': 0}                       16.6   
25  {'High': 8, 'Mid': 4, 'Low': 1, 'Very Low': 0}                       16.6   
34  {'High': 9, 'Mid': 4, 'Low': 1, 'Very Low': 0}                       16.6   
33  {'High': 9, 'Mid': 3, 'Low': 2, 'Very Low': 0}                       16.6   
24  {'High': 8, 'Mid': 3, 'Low': 2, 'Very Low': 0}                       16.6   
15  {'High': 7, 'Mid': 3, 'Low': 2, 'Very Low': 0}                       16.6   
9   {'High': 6, 'Mid': 3, 'Low': 2, 'Very Low': 0}                       16.6   
6   {'High': 5, 'Mid': 3, 'Low': 2, 'Very Low': 0}                       16.6   
3   {'High': 4, 'Mid': 3, 'Low': 2, 'Very Low': 0}                       16.6   
11  {'High': 6, 'Mid': 4, 'Low': 2, 'Very Low': 0}                       16.6   
17  {'High': 7, 'Mid': 4, 'Low': 2, 'Very Low': 0}                       16.6   
26  {'High': 8, 'Mid': 4, 'Low': 2, 'Very Low': 0}                       16.6   
35  {'High': 9, 'Mid': 4, 'Low': 2, 'Very Low': 0}                       16.6   
32  {'High': 9, 'Mid': 3, 'Low': 1, 'Very Low': 0}                       16.6   
23  {'High': 8, 'Mid': 3, 'Low': 1, 'Very Low': 0}                       16.6   
14  {'High': 7, 'Mid': 3, 'Low': 1, 'Very Low': 0}                       16.6   
8   {'High': 6, 'Mid': 3, 'Low': 1, 'Very Low': 0}                       16.6   
5   {'High': 5, 'Mid': 3, 'Low': 1, 'Very Low': 0}                       16.6   
2   {'High': 4, 'Mid': 3, 'Low': 1, 'Very Low': 0}                       16.6   
12  {'High': 6, 'Mid': 4, 'Low': 3, 'Very Low': 0}                       16.6   
18  {'High': 7, 'Mid': 4, 'Low': 3, 'Very Low': 0}                       16.6   
27  {'High': 8, 'Mid': 4, 'Low': 3, 'Very Low': 0}                       16.6   
36  {'High': 9, 'Mid': 4, 'Low': 3, 'Very Low': 0}                       16.6   
19  {'High': 7, 'Mid': 5, 'Low': 1, 'Very Low': 0}                       16.6   
28  {'High': 8, 'Mid': 5, 'Low': 1, 'Very Low': 0}                       16.6   
37  {'High': 9, 'Mid': 5, 'Low': 1, 'Very Low': 0}                       16.6   
20  {'High': 7, 'Mid': 5, 'Low': 2, 'Very Low': 0}                       16.6   
29  {'High': 8, 'Mid': 5, 'Low': 2, 'Very Low': 0}                       16.6   
38  {'High': 9, 'Mid': 5, 'Low': 2, 'Very Low': 0}                       16.6   
31  {'High': 9, 'Mid': 2, 'Low': 1, 'Very Low': 0}                       16.6   
22  {'High': 8, 'Mid': 2, 'Low': 1, 'Very Low': 0}                       16.6   
13  {'High': 7, 'Mid': 2, 'Low': 1, 'Very Low': 0}                       16.6   
7   {'High': 6, 'Mid': 2, 'Low': 1, 'Very Low': 0}                       16.6   
4   {'High': 5, 'Mid': 2, 'Low': 1, 'Very Low': 0}                       16.6   
1   {'High': 4, 'Mid': 2, 'Low': 1, 'Very Low': 0}                       16.6   
0   {'High': 3, 'Mid': 2, 'Low': 1, 'Very Low': 0}                       16.6   
21  {'High': 7, 'Mid': 5, 'Low': 3, 'Very Low': 0}                       16.6   
30  {'High': 8, 'Mid': 5, 'Low': 3, 'Very Low': 0}                       16.6   
39  {'High': 9, 'Mid': 5, 'Low': 3, 'Very Low': 0}                       16.6   
40  {'High': 9, 'Mid': 6, 'Low': 1, 'Very Low': 0}                       16.6   
41  {'High': 9, 'Mid': 6, 'Low': 2, 'Very Low': 0}                       16.6   
42  {'High': 9, 'Mid': 6, 'Low': 3, 'Very Low': 0}                       16.6   
43  {'High': 9, 'Mid': 6, 'Low': 4, 'Very Low': 0}                       16.6   

    Initial Revenue  Simulation Revenue  Reward Provided  \
10          1993048             2323892           350916   
16          1993048             2323892           352661   
25          1993048             2323892           354406   
34          1993048             2323892           356151   
33          1993048             2323892           298062   
24          1993048       

In [34]:
reward_combinations = list(reward_combinations_simulation.sort_values(by=["distance_from_zero"],ascending=True).reset_index(drop=True).loc[:7,'Reward Combination'].values)
reward_combinations

[{'High': 6, 'Mid': 4, 'Low': 1, 'Very Low': 0},
 {'High': 7, 'Mid': 4, 'Low': 1, 'Very Low': 0},
 {'High': 8, 'Mid': 4, 'Low': 1, 'Very Low': 0},
 {'High': 9, 'Mid': 4, 'Low': 1, 'Very Low': 0},
 {'High': 9, 'Mid': 3, 'Low': 2, 'Very Low': 0},
 {'High': 8, 'Mid': 3, 'Low': 2, 'Very Low': 0},
 {'High': 7, 'Mid': 3, 'Low': 2, 'Very Low': 0},
 {'High': 6, 'Mid': 3, 'Low': 2, 'Very Low': 0}]

# 4. 시뮬레이션

## 기존 데이터 분포 확인